apps/ledger-live-desktop/src/renderer/components/CustomImage/ImageCropper.tsx

```diff
@@ -0,0 +1,244 @@
+import { useDebounce } from "@ledgerhq/live-common/hooks/useDebounce";
+import { Flex, Icons } from "@ledgerhq/react-ui";
+import Cropper, { Area, CropperProps } from "react-easy-crop";
+import { createCanvas, getRadianAngle, rotateSize } from "./imageUtils";
+import { ImageCropError } from "@ledgerhq/live-common/customImage/errors";
+
+export type CropResult = ImageDimensions & ImageBase64Data;
+
+type Crop = Area;
+
+const MAX_ZOOM = 10;
+
+export type CropParams = {
+  crop: Crop;
+  rotationIncrements: number;
+  zoom: number;
+  imageUuid: string;
+};
+
+export type Props = ImageBase64Data & {
+  targetDimensions: { width: number; height: number };
+  onResult: (_: CropResult) => void;
+  onError: (_: Error) => void;
+  setCropParams: (_: CropParams) => void;
+  setLoading: (_: boolean) => void;
+  withButton?: boolean;
+  initialCropParams?: CropParams;
+};
+
+async function cropAndResizeImage(
+  image: HTMLImageElement,
+  targetDimensions: ImageDimensions,
+  pixelCrop: Crop,
+  rotationIncrements: number,
+): Promise<CropResult> {
+  if (!image.complete) {
+    await new Promise(resolve => {
+      image.onload = resolve;
+    });
+  }
+
+  const { canvas, context: ctx } = createCanvas(image);
+  if (!ctx) throw Error('Context "ctx" is null');
+
+  const rotDeg = rotationIncrements * 90;
+  const rotRad = getRadianAngle(rotationIncrements * 90);
+
+  // calculate bounding box of the rotated image
+  const { width: bBoxWidth, height: bBoxHeight } = rotateSize(
+    image.naturalWidth,
+    image.naturalHeight,
+    rotDeg,
+  );
+
+  // set canvas size to match the bounding box
+  canvas.width = bBoxWidth;
+  canvas.height = bBoxHeight;
+
+  // translate canvas context to a central location to allow rotating
+  ctx.translate(bBoxWidth / 2, bBoxHeight / 2);
+  ctx.rotate(rotRad);
+  ctx.translate(-image.naturalWidth / 2, -image.naturalHeight / 2);
+
+  // draw rotated image
+  ctx.drawImage(image, 0, 0);
+
+  // croppedAreaPixels values are bounding box relative
+  // extract the cropped image using these values
+  const data = ctx.getImageData(pixelCrop.x, pixelCrop.y, pixelCrop.width, pixelCrop.height);
+
+  // set canvas width to final desired crop size - this will clear existing context
+  canvas.width = pixelCrop.width;
+  canvas.height = pixelCrop.height;
+
+  // paste generated rotate image at the top left corner
+  ctx.putImageData(data, 0, 0);
+
+  // new canvas used to resize to the target dimensions
+  const { context: resizedCtx, canvas: resizedCanvas } = createCanvas();
+  if (resizedCtx === null) throw new Error('Context "croppedCtx" is null');
+  resizedCanvas.width = targetDimensions.width;
+  resizedCanvas.height = targetDimensions.height;
+  resizedCtx.drawImage(
+    canvas,
+    0,
+    0,
+    pixelCrop.width,
+    pixelCrop.height,
+    0,
+    0,
+    targetDimensions.width,
+    targetDimensions.height,
+  );
+
+  return {
+    imageBase64DataUri: resizedCanvas.toDataURL(),
```

Comment:
it works but there may be performance problem if you have big images. you could alternatively use `toBlob` and get an url of a blob. https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toBlob

Reply:
Ah!
Nice to know there is a more performant alternative to that. In that case I think I would keep it as it is as for that use case the image dimensions are quite restricted (400*672). In case we reuse this component for other more generic purposes I would implement the blob 👍

Reply:
However upstream of that component I'm currently loading images (with potentially big sizes) and using their base 64 URI so there I might optimise with a blob.